File: wot-dijkstra/src/main/java/org/pgpainless/wot/dijkstra/sq/ReferenceTime.java
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.wot.dijkstra.sq;

import javax.annotation.Nonnull;
import java.util.Date;

public interface ReferenceTime {

    @Nonnull Date getTimestamp();

    static ReferenceTime now() {
        final Date now = new Date();
        return new ReferenceTime() {
            @Override
            @Nonnull
            public Date getTimestamp() {
                return now;
            }
        };
    }

    static ReferenceTime timestamp(@Nonnull Date timestamp) {
        return new ReferenceTime() {
            @Override
            @Nonnull
            public Date getTimestamp() {
                return timestamp;
            }
        };
    }
}

File: wot-dijkstra/src/main/kotlin/org/pgpainless/wot/dijkstra/sq/ReferenceTime.kt
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.wot.dijkstra.sq

import java.util.*

/**
 * Reference time for Web of Trust calculations.
 */
interface ReferenceTime {

    /**
     * Timestamp as [Date].
     */
    val timestamp: Date

    companion object {

        /**
         * Create a [ReferenceTime] with a timestamp that corresponds to the current time.
         */
        @JvmStatic
        fun now(): ReferenceTime {
            return timestamp(Date())
        }

        /**
         * Create a [ReferenceTime] from the given [stamp] timestamp.
         */
        @JvmStatic
        fun timestamp(stamp: Date): ReferenceTime {
            return object: ReferenceTime {
                override val timestamp: Date
                    get() = stamp
            }
        }
    }
}